### PR TITLE
expression: fix comparison of float.

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve? 
TiDB has a precision problem when compare float values.

```sql
create table 02_float_precision (
	id int unsigned primary key auto_increment,
	a float(9,4)
);
insert into 02_float_precision (a) values (1.2);
select a from 02_float_precision where a = 1.2;
```

TiDB cannot find the inserted row.

### What is changed and how it works?
Use `abs(x - y) < eps` to check if two number may be equal. The value of `eps` is [same as MySQL](https://github.com/mysql/mysql-server/blob/8.0/sql/item_cmpfunc.cc#L794-L796)

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: expression/distsql_builtin.go [hunk 1]
```diff
@@ -1159,7 +1159,7 @@ func convertString(val []byte) (*Constant, error) {
 	return &Constant{Value: d, RetType: types.NewFieldType(mysql.TypeVarString)}, nil
 }
 
-func convertFloat(val []byte, f32 bool) (*Constant, error) {
+func convertFloat(val []byte, f32 bool, ft *tipb.FieldType) (*Constant, error) {
```
Comment (wjhuang2016) [inline expression/distsql_builtin.go]: Since we have `FieldType`, do we still need `f32 bool`？

File: expression/integration_test.go [hunk 2]
```diff
@@ -3433,6 +3433,20 @@ func (s *testIntegrationSuite) TestCompareBuiltin(c *C) {
...
+	tk.MustExec("insert into t values (1.2)")
+	result = tk.MustQuery("select * from t where a = 1.2")
+	c.Assert(len(result.Rows()), Equals, 0)
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a float(9, 4))")
+	tk.MustExec("insert into t values (1.2)")
+	tk.MustQuery("select * from t where a = 1.2").Check(testkit.Rows("1.2"))
+	result = tk.MustQuery("select * from t where a = 1.200001")
+	c.Assert(len(result.Rows()), Equals, 0)
```
Comment (wjhuang2016) [inline expression/integration_test.go]: Please add  test case in PR description

File: go.sum [hunk 3]
```diff
@@ -197,6 +197,7 @@ github.com/pingcap/tipb v0.0.0-20191120020146-6161b015e21e/go.mod h1:RtkHW8WbcNx
 github.com/pkg/errors v0.8.0/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pkg/errors v0.8.1 h1:iURUrRGxPUNPdy5/HRSm+Yj6okJ6UtLINN0Q9M4+h3I=
 github.com/pkg/errors v0.8.1/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
+github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
```
Comment (wjhuang2016) [inline go.sum]: We can remove this line?

File: types/compare.go [hunk 4]
```diff
@@ -111,6 +111,18 @@ func CompareFloat64(x, y float64) int {
...
 
+// CompareFloat64Fixed returns an integer comparing the float64 x to y.
+// It treat two float64 is equal if the difference of them less than eps.
+func CompareFloat64Fixed(x, y, eps float64) int {
+	if x < y {
+		return -1
+	} else if x == y || math.Abs(x-y) < eps {
+		return 0
+	}
+
```
Comment (wjhuang2016) [inline types/compare.go]: Check equal first,  what if `x < y && math.Abs(x-y) < eps` ?

Comment (bobotu) [reply types/compare.go]: Regardless of `x < y` or `x > y`, if `abs(x - y) < eps` we think they are equal under the error specified by `eps`.

Comment (wjhuang2016) [reply types/compare.go]: I mean that `x < y && math.Abs(x-y) < eps` will retrun -1 in this code.
should be:
if x == y || math.Abs(x-y) < eps {
return 0
}
if x < y {
return -1
}
return 1

File: expression/integration_test.go (hunk 2 above)
Comment (wjhuang2016) [reply expression/integration_test.go]:
```
create table 02_float_precision (
	id int unsigned primary key auto_increment,
	a float(9,4)
);
insert into 02_float_precision (a) values (1.2);
select a from 02_float_precision where a = 1.2;
```

File: types/compare.go [hunk 5]
```diff
@@ -111,6 +111,18 @@ func CompareFloat64(x, y float64) int {
 	return 1
 }
 
+// CompareFloat64Fixed returns an integer comparing the float64 x to y.
```
Comment (XuHuaiyu) [inline types/compare.go]: s/CompareFloat64Fixed/CompareFloat64WithEps ?

**Comment by breeswish** [comment]
I don't think MySQL actually counts the eps:

```
mysql> create table t(a double, b double, c double);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (0.1, 0.2, 0.3);
Query OK, 1 row affected (0.01 sec)

mysql> select * from t where a+b=c;
Empty set (0.00 sec)

mysql> 
```

**Comment by breeswish** [comment]
Additionally looks like this PR does not fix the problem listed in the PR body:

```
mysql> create table t(a float(9, 4));
Query OK, 0 rows affected (0.10 sec)

mysql> insert into t values (1.2);
Query OK, 1 row affected (0.02 sec)

mysql> select * from t where a = 1.2;
Empty set (0.00 sec)

mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-alpha-1132-gac8b9187b
Git Commit Hash: ac8b9187b8ee0fde00f0f7c88ed7aa9bfa4b33b2
Git Branch: pr/13619
UTC Build Time: 2019-12-23 07:44:19
GoVersion: go1.13.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by bobotu** [comment]
> 
> 
> I don't think MySQL actually counts the eps:
> 
> ```
> mysql> create table t(a double, b double, c double);
> Query OK, 0 rows affected (0.01 sec)
> 
> mysql> insert into t values (0.1, 0.2, 0.3);
> Query OK, 1 row affected (0.01 sec)
> 
> mysql> select * from t where a+b=c;
> Empty set (0.00 sec)
> 
> mysql> 
> ```

@breeswish PTAL https://github.com/mysql/mysql-server/blob/8.0/sql/item_cmpfunc.cc#L792 
If decimal of number is `DECIMAL_NOT_SPECIFIED` (your case) MySQL will use `compare_real`.

**Comment by breeswish** [comment]
Additionally I don't think the eps in compare is the *core* of the problem. Adding eps in compare cannot support index:

```
mysql> create table t(a float(9, 4), index(a));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (1.2);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+--------+
| a      |
+--------+
| 1.2000 |
+--------+
1 row in set (0.00 sec)

mysql> select * from t where a = 1.2;
Empty set (0.00 sec)

mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-alpha-1132-gac8b9187b
Git Commit Hash: ac8b9187b8ee0fde00f0f7c88ed7aa9bfa4b33b2
Git Branch: pr/13619
UTC Build Time: 2019-12-23 07:44:19
GoVersion: go1.13.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by bobotu** [comment]
@breeswish You are right. Basically, we are not fully compatible with this behavior, because this is non-lexical ordering. #13664 has the same problem. IMO, if we want to fully compatible with MySQL, we may need to change the encoding of data...

**Comment by breeswish** [comment]
@bobotu Yes and I think the encoding of the data is *the core problem*: if you change the encoding of the data instead, you will also get the issue in PR body fixed even without changing the compare function. I believe the eps in compare function in MySQL is only used to support correct precision *after several computing*.

**Pull request closed without merging**